omnigibson/transition_rules.py

```diff
@@ -20,37 +40,205 @@
 TransitionResults = namedtuple(
     "TransitionResults", ["add", "remove"], defaults=([], []))
 
+# Global dicts that will contain mappings
+REGISTERED_RULES = dict()
+
+
+class TransitionRuleAPI:
+    """
+    Monolithic class containing methods to check and execute arbitrary discrete state transitions within the simulator
+    """
+    # Maps BaseTransitionRule instances to list of valid object candidates combinations to check for transitions
+    ACTIVE_RULES = dict()
 
-class BaseFilter(metaclass=ABCMeta):
-    """Defines a filter to apply to objects."""
-    # Class global variable for maintaining cached state
-    # Maps tuple of unique filter inputs to cached output value (T / F)
-    state = None
+    # Maps BaseObject instances to dictionary mapping object states to arguments to set for that state when the
+    # object is initialized
+    _INIT_STATES = dict()
 
     @classmethod
-    def __new__(cls, *args, **kwargs):
+    def get_rule_candidates(cls, rule, objects):
         """
-        Initializes the cached state for this filter if it doesn't already exist
+        Computes valid input object candidates for transition rule @rule, if any exist
+
+        Args:
+            rule (BaseTransitionRule): Transition rule whose candidates should be computed
+            objects (list of BaseObject): List of objects that will be used to compute object candidates
+
+        Returns:
+            None or dict: None if no valid candidates are found, otherwise mapping, with keys "individual", and "group",
+                where each respective key maps to a nested dictionary mapping filter name to list of object(s)
+                satisfying that filter
         """
-        if cls.state is None:
-            cls.state = dict()
+        individual_candidates, group_candidates = rule.get_object_candidates(objects=objects)
+        n_individual_filters_satisfied, n_group_filters_satisfied = len(individual_candidates), len(group_candidates)
+        # Skip if the rule requirements are not met
+        if (rule.requires_individual_filters and n_individual_filters_satisfied != len(rule.individual_filters)) \
+                or (rule.requires_group_filters and n_group_filters_satisfied != len(rule.group_filters)):
+            return
 
-        return super(BaseFilter, cls).__new__(cls)
+        return dict(individual=individual_candidates, group=group_candidates)
 
     @classmethod
-    def update(cls):
+    def prune_active_rules(cls, objects):
```

Comment:
hmm, so what is the relationship between `prune_active_rules`, `refresh_rules` and `refresh_all_rules`? 

Is it possible for `prune_active_rules` to increase the set of active rules? Or will it strictly prune the set (i.e. reduce)?